src/mono/wasm/runtime/binding_support.js

```diff
@@ -235,18 +262,38 @@ var BindingSupportLib = {
 				var cont_obj = null;
 				// note that we do not implement promise/task roundtrip
 				// With more complexity we could recover original instance when this promise is marshaled back to C#.
-				var result = new Promise (function (resolve, reject) {
-					cont_obj = {
-						resolve: resolve,
-						reject: reject
-					};
+				var result = new Promise(function (resolve, reject) {
+					if (self._js_owned_object_registry) {
+						cont_obj = {
+							resolve: resolve,
+							reject: reject
+						};
+					} else {
+						// when FinalizationRegistry is not supported by this browser, we will do immediate cleanup after use
```

Comment:
This logic is complex enough that we may want to introduce a config flag in BINDING that controls whether the registry is being used, so that we can then exercise it in a couple automated tests. As-is, the only thing verifying these safari flows work will be people manually testing on old Safari.

My suggestion would be something like ```BINDING._enable_finalization_registry = (does this runtime environment have FinalizationRegistry)``` and then instead of checking whether the registry exists, code checks that bool. Then an automated test can explicitly set it to false in order to make the APIs it calls not use the registry and we can exercise all these code paths.